src/lib/OAuth2/OAuth2.ts

```diff
@@ -0,0 +1,193 @@
+import OAuth2UserImpl from 'lib/OAuth2/OAuth2User';
+import {
+  User,
+  UserManager,
+  UserManagerSettings,
+  WebStorageStateStore,
+} from 'oidc-client';
+
+export enum OAuth2Events {
+  UserLoaded = 'userLoaded',
+  TokenExpired = 'tokenExpired',
+  TokenExpiring = 'tokenExpiring',
+  UserUnloaded = 'userUnloaded',
+  UserSignedOut = 'userSignedOut',
+  SilentRenewError = 'silentRenewError',
+}
+
+interface IOAuth2EventCallbacks {
+  [OAuth2Events.UserLoaded]: (event: CustomEvent<OAuth2UserImpl>) => void;
+  [OAuth2Events.TokenExpired]: () => void;
+  [OAuth2Events.TokenExpiring]: () => void;
+  [OAuth2Events.UserUnloaded]: () => void;
+  [OAuth2Events.UserSignedOut]: () => void;
+  [OAuth2Events.SilentRenewError]: (error: CustomEvent<Error>) => void;
+}
+
+export interface IOAuth2Config {
+  authority: string;
+  clientId: string;
+  clientSecret: string;
+  redirectUri: string;
+  responseType?: string;
+  responseMode?: 'query' | 'fragment';
+  scope?: string;
+  prompt?: string;
+  automaticSilentRenew?: boolean;
+  includeIDTokenInSilentRenew?: boolean;
+  loadUserInfo?: boolean;
+  revokeAccessTokenOnLogout?: boolean;
+  filterProtocolClaims?: boolean;
+  validateSubOnSilentRenew?: boolean;
+  persistenceMethod?: Storage;
+}
+
+class OAuth2 {
```

Comment:
This is a wrapper around the `oidc-client` module.

This provides support for OAuth2 authentication, including persistence between sessions, token renewal and silent logins.